Remove use of lock from `OAuth2AuthorizationGrant`

ks-yim opened the pull request:
#### Motivation:
* Currently, `OAuth2AuthorizationGrant` uses a lock to guarantee that its access token gets updated atomically and `OAuth2Client` depends on `OAuth2AuthorizationGrant`.
* We may want to remove the use of lock and introduce a better way to go async.

#### Modifications:
* Remove the use of `TokenLifeCycleManager` which internally uses locks to guarantee serial execution of concurrent token refresh requests.
* Adopt CAS to atomically update `tokenFuture` which guarantees only the first token-issuing flow gets executed and others wait for it to use an updated token.
* (Breaking) `OAuth2AuthorizationGrant#withAuthorization` method is removed.

#### Result:
* `OAuth2AuthorizationGrant` no longer uses a lock.
* Closes #3571

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/OAuth2AuthorizationGrant.java:
```diff
@@ -42,12 +43,12 @@
-            return req.withHeaders(req.headers().toBuilder()
-                                   .set(HttpHeaderNames.AUTHORIZATION, accessToken.authorization())
-                                   .build());
+    default CompletionStage<HttpRequest> withAuthorization(ClientRequestContext ctx, HttpRequest req) {
```
I think users could call this method before a `ClientRequestContext` is made.
If you want to acquire an event loop, how about forking the following code?
https://github.com/line/armeria/blob/73c4e0ffcc0a544ee2b435aefd27acf973fece03/core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java#L442-L443

ks-yim replied:
Thanks for the advice!

My understanding was `XxxGrant` classes are only meant to be used in `OAuth2Client` decorator.
So I've thought we need a `ContextAwareEventLoop` as the authentication occurs within the context of armeria's client request.

Maybe I am not getting the whole picture of this implementation 😂

ikhoon replied:
> My understanding was XxxGrant classes are only meant to be used in OAuth2Client decorator.

Acutally, `OAuth2AuthorizationGrant.withAuthorization(req)` was suggested by me to inject `Authorization` header into the existing `HttpRequest`.

ks-yim replied:
Thanks for the reference, let me check the concept and update the PR.

ks-yim replied:
I am sorry for not being able to update the PR in a timely manner.
I've reverted a change in the interface as per your suggestion and the approach suggested by @minwoox removed the needs for acquiring an event loop in the end.

But, I still think being able to access `ctx` in `withAuthorization` may benefit us in a few situations.

Perhaps, we may want to introduce a new grant type in the future like `TokenPerUserXXXGrantType` where we need to load an access token for each logged in user and `ctx` will most likely be the right place where we put or get`User` objects. Then, it's just a matter of overriding `abstract User loadUser(ClientRequestContext ctx)` method.

minwoox replied:
Hm, I don't see any case that a user would call `withAuthorization` before a `RequestContext` is created. Let me talk to @ikhoon later. 😄

ikhoon replied:
I imagine that users could want to acquire an access token using an `OAuth2ClientCredentialsGrant` and send a request to another resource server.
However, I am not an expert in OAuth2. Any thoughts @max904-github?

ks-yim replied:
We may handle this in another PR once we got an answer from @max904-github!

minwoox replied:
Oops, I forgot to mention about it. 😅 Let me talk to @ikhoon.

minwoox replied:
Had a chat with @ikhoon about it.
We can hide this API `withAuthorization` because it's only used in `OAuth2Client`.
How about moving the method inside `OAuth2Client` and make it private?

ks-yim replied:
@minwoox Thanks for the suggestion!
I've moved `withAuthorization` method to `OAuth2Client` but still without modifying the method signature.
Please check out if this is what you'd expect. 🙏

minwoox replied:
How about just inlining it so that we don't create an additional future? 😄

ks-yim replied:
I shouldn't have just copy & pasted 😂.
All updated!

minwoox commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +90,101 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
-        return tokenManager.get();
+    public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken(ClientRequestContext ctx) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        doGetAccessToken(ctx, future, 1);
```
nit:
```suggestion
        getAccessToken(ctx, future, 1);
```
We can use the same name because the parameters are different.

ks-yim replied:
I've removed this method adopting your approach, thanks! 👍

minwoox commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +90,101 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+            return;
+        }
+        if (!authenticatingUpdater.compareAndSet(this, 0, 1)) {
+            scheduleNextRetry(ctx, future, attempts);
```
If this condition is false, it means that another executor is authenticating to get the access token. So I think we don't have to schedule the next retry but just use the token when it's ready.
How about changing:
```java

private static final AtomicReferenceFieldUpdater<AbstractOAuth2AuthorizationGrant, CompletableFuture> accessTokenFutureUpdater = ....

private volatile CompletableFuture<GrantedOAuth2AccessToken> accessTokenFuture;

private void doGetAccessToken(...) {

    ...    
    if (!accessTokenFutureUpdater.compareAndSet(this, null, future)) {            
        accessTokenFuture.handle((...) -> {... future.complete(token)...});
    }
    ...
}
```

minwoox reviewed (commented):
The approach looks good! 👍

minwoox commented:
@ks-yim Let us know if you need some help or are ready for the review. 😄

ks-yim commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+    private CompletableFuture<GrantedOAuth2AccessToken> refreshAccessToken(GrantedOAuth2AccessToken token) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        refreshRequest.make(token).handle((newToken, cause) -> {
+            if (cause != null) {
+                if (cause instanceof TokenRequestException) {
+                    // try to issue a new access token from scratch
+                    final CompletableFuture<GrantedOAuth2AccessToken> tokenUpdateFuture =
+                            RequestContext.mapCurrent(ctx -> ctx.makeContextAware(obtainAccessToken(token)),
+                                                      () -> issueAccessToken(token));
+                    tokenUpdateFuture.handle((newToken0, cause0) -> {
+                        if (cause0 != null) {
+                            future.completeExceptionally(cause0);
+                        } else {
+                            future.complete(newToken);
+                        }
+                        return null;
+                    });
+                } else {
+                    future.completeExceptionally(cause);
+                }
```
Would it be better to extract exception handling logic using `.exceptionally(...)`?

ks-yim commented:
> @ks-yim Let us know if you need some help or are ready for the review. 😄

@minwoox @ikhoon Thanks for being so patient with me. 😅
I've updated the PR and cleaned up unused code adopting the suggested approach. PTAL 🙏.

And I am thinking about a way to test if concurrency is well regulated with the new approach, though it passes all unit tests.

minwoox commented:
> Thanks for being so patient with me.

Oh, please take your time. 😄 
Just wanted to know if you have trouble or not. 😄 

> And I am thinking about a way to test if concurrency is well regulated with the new approach,

Let me see what I can do.

trustin commented:
@ks-yim Please feel free to let us know if it's ready for review. I'd be happy to take a look then.

minwoox commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+    }
+
+    private CompletableFuture<GrantedOAuth2AccessToken> refreshAccessToken(GrantedOAuth2AccessToken token) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
```
How about passing the `future` https://github.com/line/armeria/pull/3618/files#diff-866e837e1f45041da819b2fed992548e97c9b2a8a72071e68c374f9d4678e53aR123 instead of creating a new one?
Then, we can change here https://github.com/line/armeria/pull/3618/files#diff-866e837e1f45041da819b2fed992548e97c9b2a8a72071e68c374f9d4678e53aR123

```java
if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
    if (token != null && token.isRefreshable()) {
        refreshAccessToken(token, future);
        return null;
    }
    ...
}
```

ks-yim replied:
Nice, let me update the PR and ping you again!

ks-yim replied:
Done!

minwoox reviewed (commented):
Thanks! Left some more comments. 😉

ks-yim commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+
+        tokenFuture.handle((token, ignored) -> {
```
If `tokenFuture` is completed exceptionally, we have to initiate another token issuing flow(the `future` here) and it shouldn't complete exceptionally inheriting the `cause` of `tokenFuture`.
It will never get a chance to update `tokenFuture` if we do this.

And expected behavior in `handle` callback is the same for both of the following cases:
* token == null && cause == null
* token == null && cause != null

So I've chosen to ignore the cause.

ks-yim commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+                    future.complete(newToken);
+                    if (saveTokenFunc != null) {
+                        // suppress 'ReturnValueIgnored' warning
+                        final CompletableFuture<Void> unused = saveTokenFunc.apply(newToken);
```
I wanted to force async execution by interface so that the API become less error-prone.
I am down if you think `Consumer` interface and documentation will do the work.

ks-yim commented:
@ikhoon @minwoox @trustin It's ready for review now! Please take another look 🙏.

Here's the brief summary of the update:
* Atomically update `tokenFuture` leveraging CAS and let other requests wait for it in concurrent the token update scenario (as per @minwoox's suggestion)
* Unit tests to test concurrency in the following scenarios.
  - Concurrent initial token update
  - Concurrent token refresh
* Removed unused implementations and test cases
  - `TokenLifeCycleManager`
  - `SerialFuture`


And I'd like to ask for your advice on the following points though the PR is complete as it is now:
* Removing or improving `saveTokenFunc` and `loadTokenFunc`
  - Synchronization of token throughout multiple server processes can't properly be done without inter-process lock though I tried to follow the behavior of the past implementation with this PR.
  - Having persistency layer with minimal or no inter-process lock mechanism might still be useful in a `token-per-user` scenario but it seems current `AbstractOAuthAuthorizationGrant` implementations are intended to be used only for server authentication.

* Making `OAuthAuthorizationGrant#getAccessToken()` accept `ClientRequestContext` to provide a naturally way to access armeria context.
  - Pros
    + Doesn't lose contextual information in a series of async calls during token update.
  - Cons
    + `getAccessToken()` may get called before `ClientRequestContext` is made.

minwoox commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +98,83 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+    public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+
+        tokenFuture.handle((token, ignored) -> {
+            if (isValidToken(token)) {
+                future.complete(token);
+                return null;
+            }
+            if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+                if (token == null && loadTokenFunc != null) {
+                    loadTokenFunc.get().handle((storedToken, unused) -> {
+                        if (isValidToken(storedToken)) {
+                            future.complete(storedToken);
+                            return null;
+                        }
+                        issueAccessToken(token, future);
+                        return null;
+                    });
+                    return null;
+                }
+                if (token != null && token.isRefreshable()) {
+                    refreshAccessToken(token, future);
+                    return null;
+                }
+                issueAccessToken(token, future);
+                return null;
+            }
+            this.tokenFuture.handle((newToken, cause) -> {
+                if (cause != null) {
+                    future.completeExceptionally(cause);
+                } else {
+                    future.complete(newToken);
+                }
+                return null;
+            });
+            return null;
+        });
+        return future;
+    }
```
`tokenFuture` is initialized with an empty future which is useless.
It also makes the code slanted to the right because the logic starts in `handle` function.
So how about `tokenFuture` as `Nullable` and do this?
```suggestion
    public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
        GrantedOAuth2AccessToken token = null;
        if (tokenFuture != null && !tokenFuture.isCompletedExceptionally()) {
            token = tokenFuture.join();
            if (isValidToken(token)) {
                return tokenFuture;
            }
        }

        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
        if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
            this.tokenFuture.handle((newToken, cause) -> {
                if (cause != null) {
                    future.completeExceptionally(cause);
                } else {
                    future.complete(newToken);
                }
                return null;
            });
            return future;
        }

        if (token == null && loadTokenFunc != null) {
            loadTokenFunc.get().handle((storedToken, unused) -> {
                if (isValidToken(storedToken)) {
                    future.complete(storedToken);
                    return null;
                }
                issueAccessToken(null, future);
                return null;
            });
            return future;
        }

        if (token != null && token.isRefreshable()) {
            refreshAccessToken(token, future);
            return future;
        }
        issueAccessToken(token, future);
        return future;
    }
```

ks-yim replied:
Wouldn't this code block the thread somehow? I don't think `tokenFuture` is not guaranteed to have completed at the time we call `join()`.
```java
if (tokenFuture != null && !tokenFuture.isCompletedExceptionally()) {
    token = tokenFuture.join();
```

minwoox replied:
That's correct. 😅 
We need to do:
```java
if (tokenFuture != null && tokenFuture.isDone() && !tokenFuture.isCompletedExceptionally()) {
```

ks-yim replied:
Even after adding `tokenFuture.isDone()` and making `tokenFuture` `@Nullable`, it doesn't pass [these unit tests](https://github.com/line/armeria/pull/3618/files#diff-5e194929dbd201eac45e8aaa05dce13942127cd33ba8632df3a7921aa21549aeR177-R238) for the reason being...:
* Let's say `#1 getAccessToken()` call atomically updated `tokenFuture` and it takes 200ms for the future to complete.
* `#2 getAccessToken()` comes in at `t = 10ms` and `tokenFuture.isDone()` evaluates to `false` and it tries to update `tokenFuture` and initiate another token issuing flow.
* Updating `tokenFuture` by `#2` will still succeed even though the future updated in `#1 getAccessToken()` is still pending. Expected behavior of `#2` is to wait for the `tokenFuture`, i guess?

minwoox replied:
Ah-ha, that's correct. 😅 Let me update the code.

minwoox replied:
Could you try with this, please?
```java
public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
    final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
    GrantedOAuth2AccessToken token = null;
    if (tokenFuture != null) {
        if (!tokenFuture.isDone()) {
            return tokenFuture;
        }
        if (!tokenFuture.isCompletedExceptionally()) {
            token = tokenFuture.join();
            if (isValidToken(token)) {
                return tokenFuture;
            }
        }
    }

    final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
    if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
        return this.tokenFuture;
    }
    if (token == null && loadTokenFunc != null) {
        loadTokenFunc.get().handle((storedToken, unused) -> {
            if (isValidToken(storedToken)) {
                future.complete(storedToken);
                return null;
            }
            issueAccessToken(null, future);
            return null;
        });
        return future;
    }
    if (token != null && token.isRefreshable()) {
        refreshAccessToken(token, future);
        return future;
    }
    issueAccessToken(token, future);
    return future;
}

```

ks-yim replied:
Now I get it, thanks! This works like a charm!
Just one thing. What about making `this.tokenFuture` `@Nonnull`? This will still work and we can remove null check.

minwoox replied:
Glad that it's working now. I should've tested it before I paste the code. 😅 

> What about making this.tokenFuture @Nonnull? This will still work and we can remove null check.

The code I pasted is not an answer but just a rough idea, so please go ahead to make it better. 😄

minwoox reviewed (commented):
> Making OAuthAuthorizationGrant#getAccessToken() accept ClientRequestContext to provide a naturally way to access armeria context.

That sounds good.

> getAccessToken() may get called before ClientRequestContext is made.

I don't think it can't be called before `ClientRequestContext` is made. Could you elaborate your concern about this, please?

> Synchronization of token throughout multiple server processes can't properly be done without inter-process lock though I tried to follow the behavior of the past implementation with this PR.

Could you give us some examples of `saveTokenFunc` and `loadTokenFunc` so that we can find out if we can avoid using inter-process lock in the functions?

ks-yim commented:
@minwoox 
> I don't think it can't be called before ClientRequestContext is made. Could you elaborate your concern about this, please?

The issue has been discussion here in [this thread](https://github.com/line/armeria/pull/3618#discussion_r646370522).
Let me change the interface once it's settled.

> Synchronization of token throughout multiple server processes can't properly be done without inter-process lock though I tried to follow the behavior of the past implementation with this PR.

Imagine the following scenario that 2 servers are issuing tokens with the same credentials:
* `Server #1` issues the initial token and save it into storage.
* `Server #2` retrieves token from storage and use it as it's still valid. `Server #1` and `Server #2` are holding the same token.
* The token expires.
* `Server #1` refreshes the token and store the updated token.
* Auth server invalidates the refresh token.
* `Server #2` also tries to refresh the token but it fails to do so as the refresh token has already been invalidated.
* `Server #2` falls back to `issueAccessToken()` and start to issue a new access token from scratch.

This is what happens with both current/original implementations.

Honestly I don't see the point of having `saveTokenFunc` and `loadTokenFunc` in `AbstractOAuth2AuthorizationGrant` class as it seems this class is intended only for server-to-server authentication. We do need those functions if the application server makes calls to resource server on behalf of users, as each user session should have a reference to its access token.

ks-yim commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -41,54 +43,33 @@
-        return token.refreshToken() != null;
-    }
+    @Nullable
+    private final Consumer<? super GrantedOAuth2AccessToken> saveTokenFunc;
```
They are replacements for `tokenPersistencySupplier` and `tokenPersistencyConsumer` in the original implementation  so the names.

And I agree that it is awkward having persistent layer logic in this class and I am considering removing those completely as I've stated here(https://github.com/line/armeria/pull/3618#issuecomment-872854532).

It would be nice if I can get some advice!

ks-yim replied:
Renamed them!

minwoox commented:
> Honestly I don't see the point of having saveTokenFunc and loadTokenFunc in AbstractOAuth2AuthorizationGrant class as it seems this class is intended only for server-to-server authentication. We do need those functions if the application server makes calls to resource server on behalf of users, as each user session should have a reference to its access token.

I see. That makes sense.
We probably don't need `saveTokenFunc` and `loadTokenFunc` to load and store the token because we store the token in memory well and it doesn't cost too much to issue a new token.

However, some users still want to use the persistent layer to reuse it when the server restarts(even though it doesn't cost too much to issue a new token.) or leave a log using the function. 

So I guess it's just a matter of giving users a choice. If they want to use the functions, they can use functions. If they don't want to use functions, then simply they don't set any functions to the builder.
And I agree with @ikhoon's comment. https://github.com/line/armeria/pull/3618#discussion_r659456162 
We need to rename them. 😄

minwoox commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java:
```diff
@@ -161,45 +158,44 @@ final Duration refreshBefore() {
-    /**
-     * A {@link Consumer} to store Access Token to, to be able restore the previous session. OPTIONAL.
+     * An optional {@link Supplier} to acquire an access token before requesting it to the authorization server.
+     * If the provided {@link GrantedOAuth2AccessToken} is valid the client doesn't request a new token.
```
```suggestion
     * If the provided {@link GrantedOAuth2AccessToken} is valid, the client doesn't request a new token.
```

minwoox reviewed (approved):
Thanks! It's much cleaner now. 😄

trustin commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +98,78 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+            return this.tokenFuture;
```
Shouldn't we check again if the future is done with valid token, although it's not highly likely? You could make use of a loop:

```java

    CompletableFuture<GrantedOAuth2AccessToken> future;
    GrantedOAuth2AccessToken token = null;
    for (;;) {
        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
        if (!tokenFuture.isDone()) {
            return tokenFuture;
        }
        if (!tokenFuture.isCompletedExceptionally()) {
            token = tokenFuture.join();
            if (isValidToken(token)) {
                return tokenFuture;
            }
        }

        future = new CompletableFuture<>();
        if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
            break;
        }
    }
```

ks-yim replied:
Thanks, that's the case that I've never thought of.
Updated the PR. Now it's more robust!

trustin commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java:
```diff
@@ -161,45 +158,44 @@ final Duration refreshBefore() {
     @Nullable
-    final Executor executor() {
-        return executor;
+    public final Consumer<? super GrantedOAuth2AccessToken> newTokenConsumer() {
```
Could we hide this from the public API for consistency? Our builders don't have getters.

ks-yim replied:
Thanks for the correction, updated!

trustin reviewed (commented):
Looking nice. Just a few minor comments.

trustin commented:
@max904-github Could you check this change out? Looks like a better way to go asynchronous to me.

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +98,78 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+                future.completeExceptionally(cause);
+                return null;
+            }
+            if (newTokenConsumer != null) {
+                newTokenConsumer.accept(newToken);
+            }
+            future.complete(newToken);
```
How about wrapping `newTokenConsumer.accept()` with try-catch?
```suggestion
            if (newTokenConsumer != null) {
                try {
                newTokenConsumer.accept(newToken);
                } catch (Exception ex) {
                    logger.warn("Unexpected exception from newTokenConsumer.accept()", ex);
                }
            }
            future.complete(newToken);
```

ks-yim replied:
Fixed!

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +98,78 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+            if (cause != null) {
+                future.completeExceptionally(cause);
+            } else {
+                if (newTokenConsumer != null) {
+                    newTokenConsumer.accept(newToken);
+                }
+                future.complete(newToken);
```
Ditto, add try-catch for `newTokenConsumer.accept()`?

ks-yim replied:
Fixed!

ikhoon commented on oauth2/src/test/java/com/linecorp/armeria/internal/client/auth/oauth2/MockOAuth2ClientCredentialsService.java:
```diff
@@ -37,11 +40,19 @@
+                                       @Header("Authorization") Optional<String> auth,
+                                       @Param("grant_type") Optional<String> grantType,
+                                       @Param("scope") Optional<String> scope) {
+        final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+        // Intentionally delay execution to test concurrent token update scenario.
+        ctx.eventLoop().schedule(() -> future.complete(handleTokenGet(auth, grantType, scope)),
+                                 100L, TimeUnit.MILLISECONDS);
+        return HttpResponse.from(future);
```
nit: Use `HttpResponse.delayed()`?
```suggestion
        // Intentionally delay execution to test concurrent token update scenario.
        return HttpResponse.delayed(handleTokenGet(auth, grantType, scope), Duration.ofMillis(100);
```

ikhoon commented on oauth2/src/test/java/com/linecorp/armeria/client/auth/oauth2/OAuth2ClientCredentialsGrantTest.java:
```diff
@@ -125,13 +125,13 @@ protected void configure(ServerBuilder sb) throws Exception {
 
     @Test
-    public void testOk() throws Exception {
+    void testOk() throws Exception {
```
🙏

ks-yim commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -117,7 +98,78 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+            return this.tokenFuture;
+        }
+        if (token == null && fallbackTokenProvider != null) {
+            fallbackTokenProvider.get().handle((storedToken, unused) -> {
```
Thanks for the catch, done!

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -108,16 +93,114 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+     *
+     * <p>Renewing a token is guaranteed to be atomic even though the method is invoked by multiple threads.
+     *
+     * <p>It optionally tries to load an access token from {@link #fallbackTokenProvider}
```
Javadoc cannot link to non-public methods or fields.
```suggestion
     * <p>It optionally tries to load an access token from
     * {@link OAuth2ClientCredentialsGrantBuilder#fallbackTokenProvider(Supplier)} or 
     * {@link OAuth2ResourceOwnerPasswordCredentialsGrantBuilder#fallbackTokenProvider(Supplier)}
```

ks-yim replied:
I didn't know that, thank you for point it out!
Updated javadoc.

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java:
```diff
@@ -108,16 +93,114 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+     * which supposedly gets one by querying to a longer term storage, before it makes a request
+     * to the authorization server.
+     *
+     * <p>One may choose to provide {@link #newTokenConsumer} hook, which gets executed
```
Ditto, could you link to the builder methods?

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java:
```diff
@@ -161,45 +158,44 @@ final Duration refreshBefore() {
+     * <p>This is supposed to be used with {@link #newTokenConsumer(Consumer)} and gets executed
+     * in the following cases:
+     * <ul>
+     *     <li>Before the first attempt to acquire an access token.
+     *     <li>Before a subsequent attempt after token issue or refresh failure.
```
```suggestion
     *     <li>Before the first attempt to acquire an access token.</li>
     *     <li>Before a subsequent attempt after token issue or refresh failure.</li>
```

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java:
```diff
@@ -161,45 +158,44 @@ final Duration refreshBefore() {
+     * An optional hook which gets executed whenever a new token is issued.
+     *
+     * <p>This can be used in combination with {@link #fallbackTokenProvider(Supplier)} to store a newly issued
+     * access token which will then be retrieved by invoking {@link #fallbackTokenProvider}.
```
```suggestion
     * access token which will then be retrieved by invoking the fallback token provider.
```

ikhoon commented on oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/OAuth2Client.java:
```diff
@@ -58,14 +60,12 @@
-            }
-        }));
+        final CompletionStage<HttpResponse> future = authorizationGrant.getAccessToken().thenApply(token -> {
+            final HttpRequest newReq = req.withHeaders(req.headers().toBuilder().set(
+                    HttpHeaderNames.AUTHORIZATION, token.authorization()).build());
```
In the perspective view of code aesthetics, this does not look good. 
How about introducing a builder API that builds authorization header fluently.
For example:
```
HttpHeaders.builder()
           .authorization(token.authorization())
           .build();

```

Since it is not related to this PR, you can also handle it in a separate PR.

ks-yim replied:
Added a shortcut method for authorization header!

ikhoon reviewed (commented):
Mostly looking good!

trustin commented on core/src/main/java/com/linecorp/armeria/common/RequestHeadersBuilder.java:
```diff
@@ -126,6 +126,11 @@ default RequestHeadersBuilder acceptLanguages(LanguageRange... acceptedLanguages
+    /**
+     * Sets the <a href="https://datatracker.ietf.org/doc/html/rfc7235#section-4.2">authorization</a> header.
+     */
+    RequestHeadersBuilder authorization(String authorization);
```
Thanks! I know it doesn't really need to be a part of this PR, but how about adding `String RequestHeaderGetters.authorization()` as well? You're gonna have to:

- Add the getter method declarations in `RequestHeaderGetters` and `RequestHeaders`;
- Add the package-private getter implementation at `HttpHeadersBase` (no need to define a setter though); and
- Add the public getter implementation that calls `HttpHeadersBase.authorization()` at `DefaultRequestHeaders`.

If you're not sure, please take a look at how we implemented `RequestHeaders.method()` and `RequestHeaderGetters.method()` in `HttpHeadersBase` and `DefaultRequestHeaders`.

If you feel like this shouldn't be a part of this PR, you could just revert this and consider sending a follow-up PR.

ks-yim replied:
I will revert this and make a separate PR.
Thanks for giving me a solid guidance! 👍

trustin commented:
@ks-yim wrote:

> Removing or improving `saveTokenFunc` and `loadTokenFunc`

I concur with @minwoox's answer on this question. +1 on improving it rather than removing it.

> Making `OAuthAuthorizationGrant#getAccessToken()` accept `ClientRequestContext` to provide a naturally way to access armeria context.

It looks to me that `OAuth2AuthorizationGrant` has been designed to work even if it's not used as a part of a client call. Could you let me know what you want to do with the context object, so I can understand this problem better?

ks-yim commented:
Sorry for the late response, @trustin.
> It looks to me that OAuth2AuthorizationGrant has been designed to work even if it's not used as a part of a client call. Could you let me know what you want to do with the context object, so I can understand this problem better?

My assumption was `OAuth2AuthorizationGrant` is intended to be used only within `OAuth2Client` where it's always guaranteed to have access to `ClientRequestContext`.

The benefit of having `ClientRequestContext` in the method signature is not huge with the current implementations of `OAuth2AuthorizationGrant` other than that we can bind a `RequestContext` to chained async calls, but I thought we may benefit from it if we were to introduce additional implementations in the future.

Imagine the case where a web server application tries to access its user's resources in Google upon user actions (https://developers.google.com/identity/protocols/oauth2#webserver). Then we may want to introduce something like `TokenPerUserAuthorizationGrant` to abstract away a complex process of binding each user session to their Google access token. The class accesses `RequestContext` to extract user information and use it to retrieve the user's token from a storage.
So once properly configured with `OAuth2Client`, calling google API on behalf of a user becomes as simple as doing just `googleApiClient.execute(...)` without having to write code to retrieve the right token and passing it to authorization header.

It could be a contrived example though, as it's quite normal that FE side manages the user access token 😅.

And JFYI, we had a discussion on it in [this thread](https://github.com/line/armeria/pull/3618#discussion_r646370522).

trustin commented:
@ks-yim Sounds reasonable to me. Please feel free to add `ClientRequestContext` as a new parameter there. Are you going to make the change in this PR or in a follow-up one?

trustin reviewed (approved):
Looks great to me :slightly_smiling_face: 
<img src="https://media.giphy.com/media/Z3yHSilEYGrSfKrYrk/giphy.gif" width="160">

ks-yim commented:
Thanks, @trustin.
>  Are you going to make the change in this PR or in a follow-up one?

Let me make a follow-up PR after getting some advice from @max904-github!

trustin commented:
Great. Let's merge once @ikhoon approves.

ikhoon reviewed (approved):
Nice work! @ks-yim 
By the way, https://github.com/line/armeria/pull/3618/files#r669456222 was not pushed to this PR.

trustin commented:
> By the way, #3618 (files) was not pushed to this PR.

IIRC We chose to do it in a follow-up PR.

ikhoon commented:
> IIRC We chose to do it in a follow-up PR.

That makes sense. We don't have to do it in this PR.
